UIforETW/Utility.cpp

```
@@ -20,7 +20,305 @@ limitations under the License.
+{
+	const LONG closeKey = ::RegCloseKey(key);
+	if (closeKey == ERROR_SUCCESS)
+		return;
+	outputPrintf(L"Failed to close registry key `%s`.\n", keyName);
+	debugLastError();
+}
+
+void CloseFindHandle(_In_ _Pre_valid_ _Post_ptr_invalid_ const HANDLE handle, PCWSTR const directory)
+{
+	const BOOL findClose = ::FindClose(handle);
+	if (findClose)
+		return;
+	outputPrintf(L"FindClose (for directory: `%s`) failed.\n", directory);
+	debugLastError();
+	std::terminate();
+}
+
+std::wstring GetDocumentsFolderPath()
+{
+	//must CoTaskMemFree when done!
+	PWSTR docsPathTemp = NULL;
+	const HRESULT docsPathResult = ::SHGetKnownFolderPath(FOLDERID_Documents, KF_FLAG_NO_ALIAS, NULL, &docsPathTemp);
+	if (FAILED(docsPathResult))
+	{
+		debugPrintf(L"SHGetKnownFolderPath (for Documents) failed to retrieve the path.\n");
+		std::terminate();
+	}
+	std::wstring docsPath(docsPathTemp);
+	CoTaskMemFree(docsPathTemp);
+	return docsPath;
+}
+
+void copyWPAProfileToDocuments(const bool force)
+{
+	// First copy the WPA 8.1 startup.wpaProfile file
+	std::wstring docsPath(GetDocumentsFolderPath());
+
+	if (force)
+		outputPrintf(L"\n");
+
+	if (docsPath.empty())
+	{
+		outputPrintf( L"Failed to copy WPA profile to documents. See debugger output for details.\n" );
+		return;
+	}
+
+	const std::wstring source = docsPath + WPAStartupFileName;
+	const std::wstring destDir = docsPath + std::wstring(L"\\WPA Files");
+	const std::wstring dest = destDir + WPAStartupFileName;
+	const BOOL destinationExists = ::PathFileExistsW(dest.c_str());
+	if (force || !destinationExists)
+	{
+		const BOOL makeDirResult = ::CreateDirectoryW(destDir.c_str(), NULL);
+		if (makeDirResult == 0)
+		{
+			const DWORD lastErr = ::GetLastError();
+
+			//ERROR_ALREADY_EXISTS is reasonable.
+			if (lastErr != ERROR_ALREADY_EXISTS)
+			{
+				outputPrintf(L"Something went wrong when copying the WPA profile to the documents folder.\n");

```

Yup, I was "programming on quicksand". It's better to just let it fail. I'm removing. (should I log a debug message?)